drivers/eeprom/eeprom.c

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2019 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup   drivers_eeprom
+ * @{
+ *
+ * @file
+ * @brief     Implementation for EEPROM driver API
+ *
+ * @author    Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ * @}
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "log.h"
+#include "eeprom.h"
+#include "eeprom_driver.h"
+#include "eeprom_params.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#ifndef EEPROM_SET_BUF
+/**
+ * @brief   Stack allocated buffer to implement the set fallback function
+ */
+#define EEPROM_SET_BUF              (32)
+#endif
+
+static eeprom_off_t _sizes[EEPROM_NUMOF];
+
+/**
+ * @brief       Check if the given device is valid and initialized
+ *
+ * @param[in]   dev     The device to check
+ *
+ * @retval      0       The device is not valid
+ * @retval      1       The device is valid
+ */
+static inline int _is_dev_valid(eeprom_t dev)
+{
+    return ((dev < EEPROM_NUMOF) && (_sizes[dev])) ? 1 : 0;
+}
+
+/**
+ * @brief       Check if the given range is valid
+ *
+ * @param[in]   dev     The device the range refers to
+ * @param[in]   pos     Start of the range to check
+ * @param[in]   len     Length of the range to check
+ *
+ * @retval      0       The range is not valid
+ * @retval      1       The range is valid
+ */
+static inline int _is_range_valid(eeprom_t dev, eeprom_off_t pos, size_t len)
+{
+    /* We basically want to check for "pos + len < _sizes[dev]", but we have to
+     * be a bit careful to prevent integer overflows */
+    if (len > _sizes[dev]) {
+        return 0;
+    }
+
+    if (_sizes[dev] - len < pos) {
+        return 0;
+    }
+
+    return 1;
+}
+
+void eeprom_init(void)
+{
+    for (unsigned i = 0; i < EEPROM_NUMOF; i++) {
+        _sizes[i] = eeprom_devs[i].driver->init(eeprom_devs[i].handle,
```

Comment:
We should use
```
        assert(eeprom_devs[i].driver && eeprom_devs[i].handle)
```
here and in all other functions that use these pointers to avoid silent crashes.

Reply:
Hmmm. I just thought that some drivers might not want an handle. E.g. for a periph device, the number of device is fixed. Maybe that driver prefers to have the `handle` as `NULL` and handle the device state locally, as it is safe to statically allocate a fixed amount of memory for that specific case.

But the `assert()` for the driver is a very good idea. Thanks!

Reply:
> Hmmm. I just thought that some drivers might not want an handle.

OK, that's a good argument. If the driver insists on a valid handle, the driver has to use `assert` for the handle.

Reply:
I think we should decouple underlying device params from the common eeprom init function. See how this is done with netdev.
I would also use an `eeprom_t dev` parameter for the init function.

Reply:
So providing a `setup()` taking the parameters that and an `init()` function? I never fully understood why that was done, as it increase ROM size (both by having two functions to call, and by having two times function boilerplate instructions needed to comply with the caller convention), increases the complexity (now two functions have to be called instead of one), increases the sources for errors (the order of the function calls matters), and has no apparent benefit.

Reply:
This way you get rid of all macros from `eeprom_params.h` and let the user deal with driver params setup.
This also has the benefit of not having to modify the generic API (and params) when a new external EEPROM driver is added to RIOT.

Reply:
> This way you get rid of all macros from `eeprom_params.h`

I wouldn't call it getting rid of, as the macros would only be moved to different headers.

> This also has the benefit of not having to modify the eeprom_dev source code (and params) when a new external EEPROM driver is added to RIOT.

How about not having to add them in `eeprom_params.h`, and let `eeprom_params.h` just `#include` those. That would be pretty much the same result.

Another advantage of the current approach is that `EEPROM_NUMOF` is evaluated by the compiler. With the netdevs, the user has to count the number of network devices himself. This is definitely a task better left to computers than to humans.

Reply:
I agree with @aabadie. The EEPROM API should be free from device specific stuff. That would also imply that the API only defines the type `eeprom_dev_t` for the device list `eeprom_devs`. The device list `eeprom_devs` would be defined by the board definition.

Reply:
What about users that use external EEPROM, that is not part of the board?

Reply:
How about moving the initialization out of this API and providing a function to register an eeprom device, similar to the SAUL registry. The initialization could than be performed in auto init with a single init function for each device. It would still not overcome the issue that users have to count the number of EEPROM devices, but it would provide the strict separation of configuration parameters and common API without the overhead of two init functions per device.

Reply:
They could define such a configuration file in $(APPDIR).  According the the example in PR #12333 there would be a file `eeprom_dev_conf.h` which is included by the `eeprom` module. This file is usualy in board definition but could also be in `$APPDIR` provided that that `$(APPDIR)` the first directory in the `INCLUDES` path.

`EEPROM_NUMOF` would also evaluated during compile time.

Reply:
> provided that that `$(APPDIR)` the first directory in the `INCLUDES` path.

BTW, at some point I suggested adding `$(APPDIR)` as the first entry to be able to override the default board definition by the application. But that didn't find consensus.